torch/distributed/distributed_c10d.py

```diff
@@ -588,22 +604,48 @@ def _new_process_group_helper(world_size,
         prefix_store = PrefixStore(group_name, store)
 
         if backend == Backend.GLOO:
+            if pg_options is not None:
+                assert isinstance(pg_options, ProcessGroupGloo.Options), \
+                    "Expected pg_options argument to be of type ProcessGroupGloo.Options"
+            else:
+                pg_options = ProcessGroupGloo.Options()
+                pg_options.timeout = timeout
+
+            # If user forget to set devices, we should do it by default
+            if not pg_options.devices:
+                ifname_env = getenv("GLOO_SOCKET_IFNAME")
+                if ifname_env is not None:
+                    pg_options.devices = [ProcessGroupGloo.create_device(interface=iface) for iface in ifname_env.split(",")]
+                else:
+                    pg_options.devices = [ProcessGroupGloo.create_default_device()]
+
+                pg_options.threads = len(pg_options.devices) * 2
```

Comment:
So looks like `threads` is only set if not pg_options.devices, what happens if user sets `devices` but not threads? Should we throw error or set it for them?

Reply:
Looks like it's not too much of an issue as we expose neither threads nor device in the docs. Although, we should probably do something reasonable in any permutation of args we're given, or maybe prefix `threads` and `devices` with `_`?

Reply:
I think making them be prefixed with `_` is good as this is something we don't want to expose to the user yet, and when we construct them or user choose the construct them, they should ensure both options are set. Probably need to change the internal code to use prefix as well. 

For permutation of args, I am thinking of the following pattern:
* pg_options._device=None, pg_options._threads=default -> we generate these
* pg_options._device=None, pg_options._threads!= default -> user pass in a option with threads set but not options, error to the user
* pg_options._device!=None, pg_options._threads!=default -> legit options passed by user
* pg_options._device!=None, pg_options._threads=default -> we treat this as legit as well and it's user's responsibility to ensure it's correct.

Reply:
Awesome! This sounds good to me.